File: lecture1/radix_sort/decimal/radix_sort_decimal.go
package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"os"
	"strconv"
)

type User struct {
	ID      int
	Score   int
	Name    string
	Surname string
}

func readUsers(filename string) ([]*User, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, fmt.Errorf("ошибка открытия файла: %w", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	if _, err := reader.Read(); err != nil {
		return nil, fmt.Errorf("ошибка чтения файла: %w", err)
	}

	var users []*User

	for {
		row, err := reader.Read()
		if err == io.EOF {
			break
		}

		if err != nil {
			return nil, fmt.Errorf("ошибка чтения строки: %w", err)
		}

		if len(row) < 4 {
			continue
		}

		id, err := strconv.Atoi(row[0])
		if err != nil {
			return nil, fmt.Errorf("ошибка преобразование id в int: %w", err)
		}

		score, err := strconv.Atoi(row[1])
		if err != nil {
			return nil, fmt.Errorf("ошибка преобразования score в int: %w", err)
		}

		user := &User{
			ID:      id,
			Score:   score,
			Name:    row[2],
			Surname: row[3],
		}
		users = append(users, user)
	}

	return users, nil
}

func writeUsers(filename string, users []*User) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("ошибка создания файла: %w", err)
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	for _, user := range users {
		row := []string{
			strconv.Itoa(user.ID),
			strconv.Itoa(user.Score),
			user.Name,
			user.Surname,
		}
		if err := writer.Write(row); err != nil {
			return fmt.Errorf("ошибка записи строки: %w", err)
		}
	}

	return nil
}

func getMaxDigit(users []*User) int {
	maxScore := 0
	for _, user := range users {
		if user.Score > maxScore {
			maxScore = user.Score
		}
	}

	return len(strconv.Itoa(maxScore))
}

func radixSortDecimal(users []*User) []*User {
	if len(users) == 0 {
		return users
	}

	result := users
	maxDigit := getMaxDigit(users)
	pow := 1
	for k := 0; k < maxDigit; k++ {
		count := make([]int, 10)
		for _, user := range result {
			digit := (user.Score / pow) % 10
			count[digit]++
		}

		for i := 1; i < len(count); i++ {
			count[i] += count[i-1]
		}

		temp := make([]*User, len(result))
		for i := len(result) - 1; i >= 0; i-- {
			user := result[i]
			digit := (user.Score / pow) % 10
			pos := count[digit] - 1
			temp[pos] = user
			count[digit]--
		}
		result = temp
		pow *= 10
	}
	return result
}

func main() {
	inputFilePrefix := "../data/input/decimal/users_"
	outputFilePrefix := "../data/output/decimal/go_sorted_users_"
	fileFormat := ".csv"
	usersInFile := []int{10, 50, 100, 1000, 10000}

	for _, countUsers := range usersInFile {
		inputFile := inputFilePrefix + strconv.Itoa(countUsers) + fileFormat
		outputFile := outputFilePrefix + strconv.Itoa(countUsers) + fileFormat

		users, err := readUsers(inputFile)
		if err != nil {
			fmt.Printf("Ошибка при чтении файла %s: %v\n", inputFile, err)
			continue
		}

		fmt.Printf("Прочитано %d пользователей\n", len(users))
		sortedUsers := radixSortDecimal(users)

		if err := writeUsers(outputFile, sortedUsers); err != nil {
			fmt.Printf("Ошибка при записи в файл %s: %v\n", outputFile, err)
			continue
		}

		fmt.Printf("Отсортированные данные записаны в %s\n", outputFile)
	}
}

File: lecture1/radix_sort/decimal/radix_sort_decimal.py
# Считаваем файл, например ../data/input/decimal/users_10.csv
# Сохраням данные в структуру, например массив элементов класса User для Python и соответсвующие реализации для других ЯП (Go, C, C++)

# Как мы сортируем? Естественно по score
# пройти по всем пользователям
#     найти максимальное количество цифр в записи одного числа поля score

# Обозначим номер разряда k 
# Чтобы отделить от числа значение конкретного разряда необходимо (score // 10^k) % 10
# создать результирующий список = исходному?
# Дальше итерируемся по всем разрядам от 0 до k - 1?
#     Создаём массив для сортировки подсчётом от 0 до 9
#     итерируемся по пользователям
#         находим значени заданного разряда score
#         увеличиваем значени по этому индексу в массиве частот

#     создаём массив префиксных сумм (преобразование частотного массива)
#     создаём временный массив для хранения отсортированных по заданному разряду пользователей
#     итерируемся по пользователям в обратном порядке
#         расставляем для конкретного разряда пользователь в отсортированном порядке во временный массив

#     результирующий массив = временный массив

# Получаем ответ
# записываем его в файл, например ../data/output/decimal/py_sorted_users_10.csv и тд для соответсвующего количества записей в файле и выбранного ЯП для реализации

import csv
from typing import List, Dict, Any

class User:
    __slots__ = ("id", "score", "name", "surname")

    def __init__(self, id: int, score: int, name: str, surname: str):
        self.id = id
        self.score = score
        self.name = name
        self.surname = surname

def read_users(filename: str) -> List[User]:
    users = []
    with open(filename, "r", encoding="UTF-8") as f:
        reader = csv.reader(f)
        next(reader)
        for row in reader:
            if not row:
                continue

            user = User(
                id=int(row[0]),
                score=int(row[1]),
                name = row[2],
                surname= row[3]
            )
            users.append(user)
    return users
    
def write_users(filename: str, users: List[User]):
    with open(filename, "w", encoding="UTF-8", newline="") as f:
        writer = csv.writer(f)
        for user in users:
            writer.writerow([user.id, user.score, user.name, user.surname])

def get_max_digit(users: List[User]) -> int:
    max_score = 0
    for user in users:
        if user.score > max_score:
            max_score = user.score

    return len(str(max_score))

def radix_sort_decimal(users: List[User]) -> List[User]:
    """
    Поразрядная сортировка по score (десятичная реализация)
    Возвращает новый отсортированный список
    """

    if not users:
        return users
    
    result = users.copy()

    max_digit = get_max_digit(users)
    for k in range(max_digit):
        count = [0]*10
        for user in result:
            digit = (user.score // (10**k)) % 10
            count[digit] += 1

        for i in range(1, len(count)): # лучше вместо len(count) использовать сразу '10', НО так менее читаемо
            count[i] += count[i-1]

        temp = [None] * len(result)
        for user in reversed(result):
            digit = (user.score // (10**k)) % 10
            pos = count[digit] - 1
            temp[pos] = user
            count[digit] -= 1

        result = temp

    return result

def main():
    input_file_prefix = "../data/input/decimal/users_"
    users_in_file = [10, 50, 100, 1000, 10000]
    output_file_prefix = "../data/output/decimal/py_sorted_users_"
    file_format = ".csv"

    for users_count in users_in_file:
        input_file = input_file_prefix + str(users_count) + file_format
        output_file = output_file_prefix + str(users_count) + file_format

        users = read_users(input_file)
        print(f"Прочитано {len(users)} пользователей")
        sorted_users = radix_sort_decimal(users)
        write_users(output_file, sorted_users)
        print(f"Отсортированные данные записаны в {output_file}")


if __name__ == "__main__":
    main()


